sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,1525 @@
+
+            sample_indices.reserve(close_pair.first-open_pair.first+1);
+
+            // get _every_ sample indices between open and close
+            for (AbcIndex_t i = open_pair.first; i <= close_pair.first; i++)
+            {
+                sample_indices.push_back(i);
+            }
+        }
+        return sample_indices;
+    }
+
+    // extract an appleseed xform sequence from given alembic xform schema
+    asr::TransformSequence extract_xform_seq(
+            const AbcGeom::IXformSchema& schema,
+            const AbcChrono_t shutter_open,
+            const AbcChrono_t shutter_close,
+            const AbcChrono_t time_offset)
+    {
+        assert(schema.getNumOps() > 0);
+        assert(shutter_open <= shutter_close);
+
+        const Abc::TimeSamplingPtr time_sampling = schema.getTimeSampling();
+
+        // find samples in shutter
+        const auto sample_indices = schema_to_sample_times(
+            shutter_open  + time_offset,
+            shutter_close + time_offset,
+            schema.getNumSamples(),
+            time_sampling);
+
+        // create and return the xform sequence from sample founds.
+        asr::TransformSequence xform_seq;
+
+        for (const AbcIndex_t i : sample_indices)
+        {
+            // time
+            const AbcChrono_t t = time_sampling->getSampleTime(i);
+
+            // alembic xform -> appleseed xform
+            const auto sample = schema.getValue(i);
+            const auto abc_mtx = sample.getMatrix();
+            const auto mtx = asf::Matrix4d(abc_mtx);
+            const auto xform = asf::Transformd::from_local_to_parent(mtx);
+
+            // as all samples are time offsetted, we offset them back
+            xform_seq.set_transform(t-time_offset, xform);
+        }
+
+        return xform_seq;
+    }
+
+    // log information for given alembic archive
+    void log_archive(Abc::IArchive& archive,
+                     AbcCoreFac::IFactory::CoreType core_type)
+    {
+        RENDERER_LOG_INFO("archive name: %s\n"
+                          "        time sampling count: %d\n"
+                          "        version: %d",
+            archive.getName().c_str(),
+            archive.getNumTimeSamplings(),
+            archive.getArchiveVersion());
+        switch(core_type)
+        {
+            case AbcCoreFac::IFactory::kHDF5:
+                RENDERER_LOG_INFO("        core type: HDF5");
+                break;
+            case AbcCoreFac::IFactory::kOgawa:
+                RENDERER_LOG_INFO("        core type: Ogawa");
+                break;
+            case AbcCoreFac::IFactory::kLayer:
+                RENDERER_LOG_INFO("        core type: Layer");
+                break;
+            case AbcCoreFac::IFactory::kUnknown:
+                RENDERER_LOG_WARNING("        core type: Unknown");
+                break;
+            default:
+                break;  // this should never happen (tm)
```

Comment:
In that case, assert:

    assert(!"This should never happen.");

Or use `assert_otherwise` (defined in `foundation/utility/otherwise.h`).

Reply:
Done.